discord/ext/commands/core.py

```diff
@@ -663,10 +677,14 @@ def __str__(self):
         # first, call the command local hook:
         cog = self.cog
         if self._before_invoke is not None:
-            if cog is None:
+            try:
+                instance = self._before_invoke.__self__  
+                # should be cog unless the hook isn't in the cog
+            except AttributeError:
+                # __self__ only exists for methods, not functions
                 await self._before_invoke(ctx)
             else:
-                await self._before_invoke(cog, ctx)
+                await self._before_invoke(instance, ctx)
```

Comment:
This is actually wrong. For regular cogs, e.g.

```py
class A(commands.Cog):
    @commands.command()
    async def foo(self, ctx):
        pass

    @foo.before_invoke
    async def before_foo(self, ctx):
        pass
```

Doing something like `bot.get_cog('A').foo._before_invoke.__self__` will be an attribute error -- because it's a function and it's not bound. The change-set suggested breaks that case. This is the same with `after_invoke`.